apps/src/code-studio/assets/show.js

```diff
@@ -1,11 +1,12 @@
 /* global dashboard */
 
 import Sounds from '../../Sounds';
-var React = require('react');
-var ReactDOM = require('react-dom');
-var ImagePicker = require('../components/ImagePicker');
-var SoundPicker = require('../components/SoundPicker');
-var Dialog = require('../LegacyDialog');
+import React from 'react';
+import ReactDOM from 'react-dom';
+import loadable from 'loadable-components';
+const ImagePicker = loadable(() => import('../components/ImagePicker'));
+const SoundPicker = loadable(() => import('../components/SoundPicker'));
+import Dialog from '../LegacyDialog';
```

Comment:
My first thought looking at this was that the `Dialog` import is going to be hoisted above the `const` declarations... but then I realized I have no idea if that's true anymore.  So I'm curious exactly how `@babel/plugin-syntax-dynamic-import` works and if import hoisting isn't a thing we should assume anymore (not that it _usually_ matters, but just in case).

The [implementation file is tiny](https://github.com/babel/babel/blob/master/packages/babel-plugin-syntax-dynamic-import/src/index.js) and just references into a config setting in Babel, so it's kind of funny that it's a plugin at all.

Looks like the `dynamicImport` option was originally added in https://github.com/babel/babylon/pull/163 (the `babylon` project has since been pulled into the main `babel/babel` repo as the parser) and one of the key changes was in the statement parser.  The latest version has abstracted this change into a secondary option [`allowImportExportEverywhere`](https://github.com/babel/babel/blob/master/packages/babel-parser/src/parser/statement.js#L231-L236) but I'm still not seeing where this is set.

I did some other digging in https://github.com/babel/babel and I can't find the actual hoisting behavior in the parser, so maybe I'll just run some local experiments on this.

Reply:
I was not familiar with scope hosting, but here's what I've found:
* https://webpack.js.org/configuration/mode states that module concatenation is enabled when webpack is run in production mode
* https://webpack.js.org/plugins/module-concatenation-plugin/ states that module concatenation implies scope hoisting
* https://webpack.js.org/plugins/module-concatenation-plugin/#optimization-bailouts lists a number of scenarios where it bails on scope hoisting. none of those mention the order of `import` statements within a file. they do, however, mention that an `import()` statement creates the "Root" of a new module, which is expected and in fact illustrated by the bundle analysis in this PR's description.

> the Dialog import is going to be hoisted above the const declarations

my intuitive understanding of the word "hoist" is that it has more to do with pulling an entire module into the module that imports it, and does not have to do with moving statements around within a file. I have not found a definition which proves that, but I don't see anything in the module concatenation docs which mentions moving statements within files.

If you have a concern about correct behavior of this code please let me know a bit more about it?

Reply:
My question actually is about moving statements around, in particular about the ordering of side-effects in imported modules.  I built a [runnable example](https://github.com/code-dot-org/code-dot-org/compare/hoisting-example) of import hoisting to help clarify my thoughts:

```js
console.log('-- A --');
require('./thisFileLogsX');
console.log('-- B --');
import './thisFileLogsY';
console.log('-- C --');
```
Current behavior:
```
-- Y --
-- A --
-- X --
-- B --
-- C --
```

As Webpack is currently configured, imports are "hoisted" to the top of the module, the same way variable declarations are hoisted to the top of their scope in JavaScript.  This means any side-effects in the files they import may run before code that appears to come before the import statement. On the other hand, `require()` runs when it is called, like any other code.

In practice this is rarely an issue because we try not to use modules with side-effects, but it comes up occasionally when interfacing with third-party libraries that depend on one another via globals (e.g. jQuery and its plugins).

I assume that dynamic `import()` calls do not get hoisted - they should behave like require() does now.  My question is: Does this configuration change affect regular `import` statements too? Will it affect the behavior of this example?



If you run this test file with `--entry` you can see the output yourself.

Reply:
Interesting note on definitions -- the [webpack docs](https://webpack.js.org/plugins/module-concatenation-plugin/) call "module concatenation" and "scope hoisting" the same thing, and [blog posts about webpack scope hoisting](https://medium.com/webpack/brief-introduction-to-scope-hoisting-in-webpack-8435084c171f) describe it as pulling content out of a child module up into the parent module which requires it. However, [MDN docs](https://developer.mozilla.org/en-US/docs/Glossary/Hoisting) and [blog posts about javascript hoisting](https://medium.com/backticks-tildes/understanding-hoisting-and-scoping-in-javascript-39ea7b41e31) describe "hoisting" as moving certain statements to the top of the file. So, you used the term "hoisting" correctly and I mistook it for webpack's "scope hoisting".

The good news is I've run your example and I see the same order:
```
Dave-MBP:~/src/cdo/apps (lazy-load-js)$ grunt karma:entry --entry=./test/unit/hoist/hoistingTest.js 
...
PhantomJS 2.1.1 (Mac OS X 0.0.0) LOG: '-- Y --'
PhantomJS 2.1.1 (Mac OS X 0.0.0) LOG: '-- A --'
PhantomJS 2.1.1 (Mac OS X 0.0.0) LOG: '-- X --'
PhantomJS 2.1.1 (Mac OS X 0.0.0) LOG: '-- B --'
PhantomJS 2.1.1 (Mac OS X 0.0.0) LOG: '-- C --'
LOG: 'The test ran.'
```

Reply:
:+1: LGTM then!  It makes sense, given dynamic imports have a different syntax.  Just wanted to make sure of no surprises here.